### PR TITLE
RD-726 Gap detection preprocessing

### PR DESCRIPTION
**Pull request by BlayneJLees**
# RD 726/gap detection preprocessing #

This document is a template for the pull requests of the R&D team. Please fill
the different parts, and remove the current instructions, before submitting the
PR.

## Description ##

Repo up until preprocessing 
Jira ticket : [RD-726]()

## Checklist ##

- [x] PEP8 compliant (run `black --check $(git diff --name-only HEAD origin/develop | grep .py) -l 119`);
- [ ] Passing tests;
- [ ] _(If required)_ Confluence pages updated.

### CODE REVIEW

**Comment by BlayneJLees** [comment]
@aj-hitchins line creator is at the stage you sent it to me, as most of the work done has been back in notebooks.

File: hummingbirdtech/gap_detection/algorithm/preprocessing.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        :survey: hummingbirdtech.domain.survey.repository
+        :rtype:  list[shapely.geometry.linestring.LineString]
+        """
+        image = self._rgb_image_store.load(survey)
+        # return only the row band
+        row_band = self._wide_row_segmenter.segements(image=image)[0]
+
+        # normalise the band between 0 and 1
+        transformed_row_band = row_band.transform(
+            lambda array: np.where(array == 255, 1, 0)
```
Comment (aj-hitchins) [inline hummingbirdtech/gap_detection/algorithm/preprocessing.py]: Not sure if normalise is the right word here? This fn will just replace every pixel with the value `255` with the value `1`, and everything `not 255` with `0`

Comment (BlayneJLees) [reply hummingbirdtech/gap_detection/algorithm/preprocessing.py]: Yes your right, also this is incorrect as the `0` and `1` should be reversed as that is the interrow band not the crop row band.

**Comment by aj-hitchins** [comment]
@BlayneJLees There seems to be the work from line reconstruction in the same branch here?

File: hummingbirdtech/gap_detection/algorithm/preprocessing.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+            local_peaks = peak_local_max(
+                block.content, min_distance=self._minimum_distance
+            )
+            for peak in local_peaks:
+                peaks.append(
+                    self._get_offsetted_peaks(x=peak[1], y=peak[0], block=block)
+                )
+
+        geographic_peaks = [
+            self._coordinate_converter.pixel_to_projected_geometry(
```
Comment (aj-hitchins) [inline hummingbirdtech/gap_detection/algorithm/preprocessing.py]: You might need to offset the point by 0.5 * pixel size to get the coordinate in the centre of the pixel:
```python
from shapely.ops import transform
point = transform(
    lambda x, y, z=None: (x+(0.5 * band.geotransform.pixel_width), y+(0.5 * band.geotransform.pixel_height)), 
    point
)

**Comment by BlayneJLees** [comment]
I had intended to split this into much smaller commits, but as I had to move both gap detection and line reconstruction into planting lines the files were detected as new files. This meant I could not split my commits into much smaller ones. I had to do this to try and ensure that the repo follows convention and to make the setup easier.

File: hummingbirdtech/planting_lines/line_reconstruction/algorithm/line_creator.py [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+
+        for coords in snapped_coords:
+            if len(coords) <= self._minimum_coordinate_length:
+                pass
+            else:
+                coords = np.array(coords)
+
+                distance = np.cumsum(np.sqrt(np.sum(np.diff(coords, axis=0) ** 2, axis=1)))
+                distance = np.insert(distance, 0, 0) / distance[-1]
+                alpha = np.linspace(0, 1, 200)
```
Comment (BlayneJLees) [inline hummingbirdtech/planting_lines/line_reconstruction/algorithm/line_creator.py]: Should all these variables be put in the parameters?

File: hummingbirdtech/planting_lines/line_reconstruction/algorithm/line_creator.py [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+                )[1:]
+            )
+        return block_lines
+
+    def _line_list_from_labels(self, labelled_pixels, block):
+        labels = np.unique(labelled_pixels)
+        line_list = []
+        for line_label in labels:
+            subset = np.where(labelled_pixels == line_label)
+            line = self.linestring_from_array(array=subset, block=block)
```
Comment (BlayneJLees) [inline hummingbirdtech/planting_lines/line_reconstruction/algorithm/line_creator.py]: Think here can do a linear regression directly on the array coords tbh.

**Comment by aj-hitchins** [comment]
Hey - looks like you've moved all the repos into `planting_lines`? I felt that this repo could contain two submodules: `line_reconstruction` and `gap_detection`. Doesn't need to all be inside planting lines.

**Comment by BlayneJLees** [comment]
@aj-hitchins Yes, I moved it all as I was unsure of how things would work setup wise. I wanted to have one version etc for the repository - happen to restructure how it needs to be!

**Comment by aj-hitchins** [comment]
@BlayneJLees Let's worry about that afterwards we can change the setup next week!

File: hummingbirdtech/planting_lines/gap_detection/algorithm/gap_detector.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+            multipolygon = MultiPolygon(candidate_polygons).buffer(0)
+            difference_lines.append(line.diffence(multipolygon))
+
+        flattened_gap_lines = [
+            line for multilinestring in difference_lines for line in multilinestring
+        ]
+        gap_lines = self._remove_duplicate_geometries(flattened_gap_lines)
+        return gap_lines
+
+    def _remove_duplicate_geometries(self, geometries):
```
Comment (BlayneJLees) [inline hummingbirdtech/planting_lines/gap_detection/algorithm/gap_detector.py]: I think this isnt best not only in in this file but also when to drop duplicates on the merged lines. Again I will not implement this until you have done the refactor, so not too many changes are occurring at once.

File: hummingbirdtech/planting_lines/gap_detection/algorithm/gap_detector.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        :type contour_levels: int
+        """
+        self._gap_detection_preprocessor = gap_detection_preprocessor
+        self._line_creator = line_creator
+        self._contour_collection = contour_collection
+        self._contour_threshold = contour_threshold
+        self._contour_levels = contour_levels
+        self._contour_value = contour_value
+
+    def get_gap_lines(self, band, crop_row_lines=None):
```
Comment (BlayneJLees) [inline hummingbirdtech/planting_lines/gap_detection/algorithm/gap_detector.py]: I have put `crop_row_lines` as an argument as i'm unsure of how we will deal with input client crop lines.

File: hummingbirdtech/planting_lines/line_reconstruction/algorithm/line_creator.py (hunk 3 above)
Comment (aj-hitchins) [reply hummingbirdtech/planting_lines/line_reconstruction/algorithm/line_creator.py]: We discussed this one.

**Comment by aj-hitchins** [comment]
Work for this PR was completed in #4 . PR will be closed.

**Pull request closed without merging**